src/lsm/forest.zig

```diff
@@ -236,76 +236,71 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
         scan_buffer_pool: ScanBufferPool,
 
         pub fn init(
+            forest: *Forest,
             allocator: mem.Allocator,
             grid: *Grid,
             options: Options,
             // (e.g.) .{ .transfers = .{ .cache_entries_max = 128, … }, .accounts = … }
             grooves_options: GroovesOptions,
-        ) !Forest {
+        ) !void {
             assert(options.compaction_block_count >= Options.compaction_block_count_min);
 
-            // NodePool must be allocated to pass in a stable address for the Grooves.
-            const node_pool = try allocator.create(NodePool);
-            errdefer allocator.destroy(node_pool);
+            forest.* = .{
+                .grid = grid,
+                .grooves = undefined,
+                .node_pool = undefined,
+                .manifest_log = undefined,
+                .compaction_pipeline = undefined,
+                .scan_buffer_pool = undefined,
+            };
 
             // TODO: look into using lsm_table_size_max for the node_count.
-            node_pool.* = try NodePool.init(allocator, options.node_count);
-            errdefer node_pool.deinit(allocator);
+            try forest.node_pool.init(allocator, options.node_count);
+            errdefer forest.node_pool.deinit(allocator);
 
-            var manifest_log = try ManifestLog.init(allocator, grid, .{
+            try forest.manifest_log.init(allocator, grid, .{
                 .tree_id_min = tree_id_range.min,
                 .tree_id_max = tree_id_range.max,
                 // TODO Make this a runtime argument (from the CLI, derived from storage-size-max if
                 // possible).
                 .forest_table_count_max = table_count_max,
             });
-            errdefer manifest_log.deinit(allocator);
+            errdefer forest.manifest_log.deinit(allocator);
 
-            var grooves: Grooves = undefined;
             var grooves_initialized: usize = 0;
-
             errdefer inline for (std.meta.fields(Grooves), 0..) |field, field_index| {
                 if (grooves_initialized >= field_index + 1) {
-                    @field(grooves, field.name).deinit(allocator);
+                    const Groove = field.type;
+                    const groove: *Groove = &@field(forest.grooves, field.name);
+                    groove.deinit(allocator);
```

Comment:
Reading `@field(...).function(...)` isn't that bad. The main reason for refactoring (here and in `Forest` as well) is to make clear each field's type since it's not trivial to understand what's going on inside a loop with too many metaprogramming instructions.